lib/route/mdb.c

```diff
@@ -0,0 +1,289 @@
+ * @arg sock		Netlink socket.
+ * @arg result		Pointer to store resulting cache.
+ * @arg flags		Flags to apply to cache before filling
+ *
+ * Allocates a new MDB cache, initializes it properly and updates it
+ * to include all Multicast Database entries currently configured in the kernel.
+ *
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_mdb_alloc_cache_flags(struct nl_sock *sock, struct nl_cache **result,
+				 unsigned int flags)
+{
+	struct nl_cache * cache;
+	int err;
+
+	cache = nl_cache_alloc(&rtnl_mdb_ops);
+	if (!cache)
+		return -NLE_NOMEM;
+
+	nl_cache_set_flags(cache, flags);
+
+	if (sock && (err = nl_cache_refill(sock, cache)) < 0) {
+		nl_cache_free(cache);
+		return err;
+	}
+
+	*result = cache;
+	return 0;
+}
+/** @} */
+
+/**
+ * @name Attributes
+ * @{
+ */
+uint32_t rtnl_mdb_get_ifindex(struct rtnl_mdb *mdb)
+{
+	return mdb->ifindex;
+}
+
+uint8_t rtnl_mdb_get_family(struct rtnl_mdb *mdb)
+{
+	return mdb->family;
+}
+
+void rtnl_mdb_add_entry(struct rtnl_mdb *mdb, struct rtnl_mdb_entry *entry) {
+  nl_list_add_tail(&entry->mdb_list ,&mdb->mdb_entry_list);
+}
+
+void rtnl_mdb_foreach_entry(struct rtnl_mdb *mdb,
+                            void (*cb)(struct rtnl_mdb_entry *, void *),
+                            void *arg ) {
+  struct rtnl_mdb_entry * entry;
+  nl_list_for_each_entry(entry, &mdb->mdb_entry_list, mdb_list) {
+			cb(entry, arg);
+		}
+}
+
+int rtnl_mdb_entry_get_ifindex(struct rtnl_mdb_entry *mdb_entry) {
+  return mdb_entry->ifindex;
+}
+
+int rtnl_mdb_entry_get_vid(struct rtnl_mdb_entry *mdb_entry) {
+  return mdb_entry->vid;
+}
+
+int rtnl_mdb_entry_get_state(struct rtnl_mdb_entry *mdb_entry) {
+  return mdb_entry->state;
+}
+
+struct nl_addr * rtnl_mdb_entry_get_addr(struct rtnl_mdb_entry *mdb_entry) {
+  return mdb_entry->addr;
+}
+
+uint16_t rtnl_mdb_entry_get_proto(struct rtnl_mdb_entry *mdb_entry) {
+  return mdb_entry->proto;
+}
+/** @} */
+
+static struct nl_object_ops mdb_obj_ops = {
+	.oo_name		= "route/mdb",
+	.oo_size		= sizeof(struct rtnl_mdb), // FIX ME
+  .oo_constructor = mdb_constructor,
+	.oo_dump = {
+	    [NL_DUMP_LINE] 	= mdb_dump_line,
+	    [NL_DUMP_DETAILS]	= mdb_dump_details,
+	},
+};
+
+struct rtnl_mdb *rtnl_mdb_alloc(void)
+{
+  return (struct rtnl_mdb *) nl_object_alloc(&mdb_obj_ops);
+}
+
+struct rtnl_mdb_entry *rtnl_mdb_entry_alloc(void)
+{
+  struct rtnl_mdb_entry *_entry;
+  
+  _entry = calloc(1, sizeof(*_entry));
+  return _entry;  
+}
+
+static struct nl_af_group mdb_groups[] = {
+	{ AF_BRIDGE,	RTNLGRP_MDB },
+	{ END_OF_GROUP_LIST },
+};
+
+static struct nl_cache_ops rtnl_mdb_ops = {
+	.co_name		= "route/mdb",
+	.co_hdrsize		= 10, // FIX ME
+	.co_msgtypes		= {
+					{ RTM_NEWMDB, NL_ACT_NEW, "new"},
+					{ RTM_DELMDB, NL_ACT_DEL, "del"},
+					{ RTM_GETMDB, NL_ACT_GET, "get"},
+					END_OF_MSGTYPES_LIST,
+				  },
+	.co_protocol		= NETLINK_ROUTE,
+  .co_groups		= mdb_groups,
+  .co_request_update	= mdb_request_update,
+  .co_msg_parser = mdb_msg_parser,
```

Comment:
indentation.

In the whole patch, please be mindful that. No trailing whitespace, and indent so that it looks good with 8 space tabs.